src/main/java/com/artipie/rpm/meta/XmlEvent.java

```diff
@@ -89,7 +123,56 @@ public void add(final Package.Meta meta) throws IOException {
                     this.writer.add(events.createCharacters(entry.content()));
                     this.writer.add(events.createEndElement("", "", tag));
                 }
-                this.writer.add(events.createEndElement("", "", pkg));
+                this.writer.add(events.createEndElement("", "", "package"));
+            } catch (final XMLStreamException err) {
+                throw new IOException(err);
+            }
+        }
+    }
+
+    /**
+     * Implementation of {@link XmlEvent} to build event for {@link XmlPackage#FILELISTS} package.
+     * @since 1.5
+     */
+    class Filelists implements XmlEvent {
+
+        /**
+         * Where to write the event.
+         */
+        private final XMLEventWriter writer;
+
+        /**
+         * Ctor.
+         * @param writer Writer to write the event
+         */
+        public Filelists(final XMLEventWriter writer) {
+            this.writer = writer;
+        }
+
+        @Override
+        public void add(final Package.Meta meta) throws IOException {
+            final XMLEventFactory events = XMLEventFactory.newFactory();
+            final HeaderTags tags = new HeaderTags(meta);
+            try {
+                new PackageAndVersion(this.writer).add(meta);
+                final String[] files = tags.baseNames().toArray(new String[0]);
+                final String[] dirs = tags.dirNames().toArray(new String[0]);
+                final Set<String> dirset = Arrays.stream(dirs).collect(Collectors.toSet());
+                final int[] did = tags.dirIndexes();
+                for (int idx = 0; idx < files.length; idx += 1) {
+                    final String fle = files[idx];
+                    if (fle.isEmpty() || fle.charAt(0) == '.') {
+                        continue;
```

Comment:
It seems this condition is never reached in tests, I'd suggest covering it with a unit test, since it's not obvious.